prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,408 @@
+ */
+static inline ssize_t efa_post_recv(struct efa_base_ep *base_ep, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_mr *efa_mr;
+	struct efa_qp *qp = base_ep->qp;
+	struct ibv_recv_wr *bad_wr;
+	struct ibv_recv_wr *wr;
+	uintptr_t addr;
+	ssize_t err, post_recv_err;
+	size_t i, wr_index = base_ep->efa_recv_wr_index;
+
+	dump_msg(msg, "recv");
+
+	err = efa_post_recv_validate(base_ep, msg);
+	if (OFI_UNLIKELY(err))
+		goto out_err;
+
+	wr = &base_ep->efa_recv_wr_vec[wr_index].wr;
+	wr->wr_id = (uintptr_t)msg->context;
+	wr->num_sge = msg->iov_count;
+	wr->sg_list = base_ep->efa_recv_wr_vec[wr_index].sge;
+
+	for (i = 0; i < msg->iov_count; i++) {
+		addr = (uintptr_t)msg->msg_iov[i].iov_base;
+
+		/* Set RX buffer desc from SGE */
+		wr->sg_list[i].length = msg->msg_iov[i].iov_len;
+		assert(msg->desc[i]);
+		efa_mr = (struct efa_mr *)msg->desc[i];
+		wr->sg_list[i].lkey = efa_mr->ibv_mr->lkey;
+		wr->sg_list[i].addr = addr;
+	}
+
+	base_ep->efa_recv_wr_vec[wr_index].wr.next = NULL;
+	if (wr_index > 0)
+		base_ep->efa_recv_wr_vec[wr_index - 1].wr.next = &base_ep->efa_recv_wr_vec[wr_index].wr;
+
+	wr_index++;
+
+	if (flags & FI_MORE)
+		return 0;
+
+	err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+	if (OFI_UNLIKELY(err)) {
+		/* On failure, ibv_post_recv() return positive errno.
+		 * Meanwhile, this function return a negative errno.
+		 * So, we do the conversion here.
+		 */
+		err = (err == ENOMEM) ? -FI_EAGAIN : -err;
+	}
+
+	base_ep->efa_recv_wr_index = 0;
+
+	return err;
+
+out_err:
+	if (wr_index > 0) {
+		post_recv_err = ibv_post_recv(qp->ibv_qp, &base_ep->efa_recv_wr_vec[0].wr, &bad_wr);
+		if (post_recv_err) {
+			EFA_WARN(FI_LOG_EP_DATA,
+				 "Encountered error %ld when ibv_post_recv on error handling path\n",
+				 post_recv_err);
+		}
+	}
+
+	return err;
+}
+
+static ssize_t efa_ep_recvmsg(struct fid_ep *ep_fid, const struct fi_msg *msg, uint64_t flags)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+
+	return efa_post_recv(base_ep, msg, flags);
+}
+
+static ssize_t efa_ep_recv(struct fid_ep *ep_fid, void *buf, size_t len,
+			   void *desc, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct iovec iov;
+	struct fi_msg msg;
+
+	EFA_SETUP_IOV(iov, buf, len);
+	EFA_SETUP_MSG(msg, &iov, &desc, 1, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static ssize_t efa_ep_recvv(struct fid_ep *ep_fid, const struct iovec *iov, void **desc,
+			    size_t count, fi_addr_t src_addr, void *context)
+{
+	struct efa_base_ep *base_ep = container_of(ep_fid, struct efa_base_ep, util_ep.ep_fid);
+	struct fi_msg msg;
+
+	EFA_SETUP_MSG(msg, iov, desc, count, src_addr, context, 0);
+
+	return efa_post_recv(base_ep, &msg, 0);
+}
+
+static inline ssize_t efa_post_send_validate(struct efa_base_ep *base_ep, const struct fi_msg *msg,
```

Comment:
I prefer to make all of these validation as assertion so we don't have to check them explicitly for non-debug build